samza-api/src/main/java/org/apache/samza/system/StreamSpec.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.system;
+
+import java.util.Properties;
+
+
+/**
+ * StreamSpec is a blueprint for creating, validating, or simply describing a stream in the runtime environment.
+ *
+ * It has specific attributes for common behaviors that samza uses.
+ *
+ * It also includes a set of properties which may be systemName-specific.
+ */
+public class StreamSpec {
```

Comment:
I see how this fits nicely with the current supported streams, but what would the system name and physical name be for an array based stream (for example)? Also, in a world where we support different types of streams wouldn't we need to know how to construct that stream (maybe that is what is meant by system name now?). Maybe we're not using StreamSpecs for arrays, etc.? If not, what is the alternative?